pandas/tseries/common.py

```diff
@@ -272,38 +272,33 @@ def convert_to_pydatetime(x, axis):
 
         return x
 
-    typs = get_dtype_kinds(to_concat)
+    if typs is None:
+        typs = get_dtype_kinds(to_concat)
 
-    # datetimetz
-    if 'datetimetz' in typs:
-
-        # if to_concat have 'datetime' or 'object'
-        # then we need to coerce to object
-        if 'datetime' in typs or 'object' in typs:
-            to_concat = [convert_to_pydatetime(x, axis) for x in to_concat]
-            return np.concatenate(to_concat, axis=axis)
+    # must be single dtype
+    if len(typs) == 1:
 
-        # we require ALL of the same tz for datetimetz
-        tzs = set([getattr(x, 'tz', None) for x in to_concat]) - set([None])
-        if len(tzs) == 1:
-            return DatetimeIndex(np.concatenate([x.tz_localize(None).asi8
-                                                 for x in to_concat]),
-                                 tz=list(tzs)[0])
+        if 'datetimetz' in typs:
+            # datetime with no tz should be stored as "datetime" in typs,
+            # thus no need to care
 
-    # single dtype
-    if len(typs) == 1:
+            # we require ALL of the same tz for datetimetz
+            tzs = set([x.tz for x in to_concat])
+            if len(tzs) == 1:
+                return DatetimeIndex(np.concatenate([x.tz_localize(None).asi8
+                                                     for x in to_concat]),
+                                     tz=list(tzs)[0])
 
-        if not len(typs - set(['datetime'])):
+        elif 'datetime' in typs:
             new_values = np.concatenate([x.view(np.int64) for x in to_concat],
                                         axis=axis)
             return new_values.view(_NS_DTYPE)
 
-        elif not len(typs - set(['timedelta'])):
+        elif 'timedelta' in typs:
             new_values = np.concatenate([x.view(np.int64) for x in to_concat],
                                         axis=axis)
             return new_values.view(_TD_DTYPE)
 
     # need to coerce to object
     to_concat = [convert_to_pydatetime(x, axis) for x in to_concat]
```

Comment:
looking at this now, I suspect this is a bug here, which I believe you have in another issue. Eg. a timedelta and a datetime (with or w/o tz) may be conveted to object, BUT they go thru the ``convert_to_datetime`` which is incorrect for timedelta.